audio/out/ao_avfoundation.m

```diff
@@ -0,0 +1,349 @@
+
+    int64_t cur_time_av = CMTimeGetNanoseconds([p->synchronizer currentTime]);
+    int64_t cur_time_mp = mp_time_ns();
+    p->end_time_av = MPMAX(p->end_time_av, cur_time_av);
+
+    int request_sample_count = ao->samplerate / 10; // 1/10 samplerate may be a good number
+    int buffer_size = request_sample_count * ao->sstride;
+    if (CMBlockBufferCreateWithMemoryBlock(
+        kCFAllocatorDefault,
+        NULL,
+        buffer_size,
+        NULL,
+        NULL,
+        0,
+        buffer_size,
+        kCMBlockBufferAssureMemoryNowFlag,
+        &block_buffer
+    ) != noErr) {
+        MP_FATAL(ao, "failed to create CMBlockBuffer\n");
+        goto error;
+    }
+    void* data[1];
+    if (CMBlockBufferGetDataPointer(block_buffer, 0, NULL, NULL, (char**)data) != noErr) {
+        MP_FATAL(ao, "CMBlockBufferGetDataPointer failed\n");
+        goto error;
+    }
+
+    int64_t delta = CMTimeGetNanoseconds(CMTimeMake(request_sample_count, ao->samplerate));
+    int real_sample_count = ao_read_data_nonblocking(ao, data, request_sample_count, p->end_time_av - cur_time_av + cur_time_mp + delta);
+    if (real_sample_count == 0) {
+        mp_sleep_ns(1000000); // avoid spinning
+        goto finish;
+    }
+
+    CMSampleTimingInfo sample_timing_into[] = {(CMSampleTimingInfo) {
+        .duration = CMTimeMake(1, ao->samplerate),
+        .presentationTimeStamp = CMTimeFromNanoseconds(p->end_time_av),
+        .decodeTimeStamp = kCMTimeInvalid
+    }};
+    size_t sample_size_array[] = {ao->sstride};
+    if (CMSampleBufferCreateReady(
+        kCFAllocatorDefault,
+        block_buffer,
+        p->format_description,
+        real_sample_count,
+        1,
+        sample_timing_into,
+        1,
+        sample_size_array,
+        &sample_buffer
+    ) != noErr) {
+        MP_FATAL(ao, "failed to create CMSampleBuffe\n");
+        goto error;
+    }
+
+    [p->renderer enqueueSampleBuffer:sample_buffer];
+
+    delta = CMTimeGetNanoseconds(CMTimeMake(real_sample_count, ao->samplerate));
+    p->end_time_av += delta;
+
+    goto finish;
+
+error:
+    ao_request_reload(ao);
+finish:
+    if (block_buffer) CFRelease(block_buffer);
+    if (sample_buffer) CFRelease(sample_buffer);
+}
+
+static void start(struct ao *ao)
+{
+    struct priv *p = ao->priv;
+
+    p->end_time_av = -1;
+    [p->synchronizer setRate:1];
+    [p->renderer requestMediaDataWhenReadyOnQueue:p->queue usingBlock:^{
+        feed(ao);
+    }];
+}
+
+static void stop(struct ao *ao)
+{
+    struct priv *p = ao->priv;
+
+    dispatch_sync(p->queue, ^{
+        [p->renderer stopRequestingMediaData];
+        [p->renderer flush];
+        [p->synchronizer setRate:0];
+    });
+}
+
+static int control(struct ao *ao, enum aocontrol cmd, void *arg)
+{
+    struct priv *p = ao->priv;
+
+    switch (cmd) {
+    case AOCONTROL_GET_MUTE:
+        *(bool*)arg = [p->renderer isMuted];
+        return CONTROL_OK;
+
+    case AOCONTROL_GET_VOLUME:
```

Comment:
the newlines between the 'case's can be removed, to be consistent with all the other code.